collector/mdadm.go

```diff
@@ -0,0 +1,200 @@
+	var currentMd string
+
+	// each md has at least the deviceline, statusline and one empty line afterwards
+	// so we will have probably something of the order len(lines)/3 devices
+	// so we use that for preallocation
+	estimateMds := len(lines) / 3
+	mdstates := make([]mdstatus, 0, estimateMds)
+
+	for i, l := range lines {
+		if !(len(l) > 0) {
+			// skip entirely empty lines
+			continue
+		}
+
+		if string(l[0]) == " " {
+			// those lines are not the beginning of a md-section
+			continue
+		} else {
+
+			if strings.HasPrefix(l, "Personalities") || strings.HasPrefix(l, "unused") {
+				// we aren't interested in lines with general info
+				continue
+			}
+
+			currentMd = strings.Split(l, " ")[0]         // name of the md-device we are parsing atm
+			isActive := !strings.Contains(l, "inactive") // activity-status of said md-device
+
+			active, total, size := evalStatusline(lines[i+1]) // parse statusline, always present
+
+			// now get the number of synced blocks
+			var syncedBlocks int64
+
+			// get the line number of the syncing-line
+			var j int
+			if strings.Contains(lines[i+2], "bitmap") { // then skip the bitmap line
+				j = i + 3
+			} else {
+				j = i + 2
+			}
+
+			// if device is syncing atm, get the number of currently synced bytes,
+			// otherwise that number equals the size of the device
+			if strings.Contains(lines[j], "recovery") || strings.Contains(lines[j], "resync") {
+				syncedBlocks = evalBuildline(lines[j])
+			} else {
+				syncedBlocks = size
+			}
+
+			mdstates = append(mdstates, mdstatus{currentMd, isActive, active, total, size, syncedBlocks})
+
+		}
+
+	}
+
+	return mdstates, nil
+
+}
+
+// just returns th epointer to a struct holding all exportable metrics
+func NewMdadmCollector() (Collector, error) {
+
+	return &mdadmCollector{
+		isActive: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "md_is_active",
+				Help: "indicator whether the md-device is active or not",
+			},
+			[]string{"device"}),
+		disksActive: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "md_disks_active",
+				Help: "number of active disks of device",
+			},
+			[]string{"device"}),
+		disksTotal: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "md_disks_total",
+				Help: "total number of disks of device",
+			},
+			[]string{"device"}),
+		blocksTotal: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "md_blocks_total",
+				Help: "total number of blocks on device",
+			},
+			[]string{"device"}),
+		blocksSynced: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "md_blocks_synced",
+				Help: "number of blocks synced on device",
+			},
+			[]string{"device"}),
+	}, nil
+}
+
+func (c *mdadmCollector) Update(ch chan<- prometheus.Metric) (err error) {
+
+	// first parse mdstat-file...
+	mdstate, err := parseMdstat(statusfile)
+	if err != nil {
+		return err
+	}
+
+	// ... and then plug the result into the metrics to be exported
+	for _, mds := range mdstate {
```

Comment:
Yeah, it's probably better to use throwaway ConstMetrics for this that are built on every scrape and then discarded. That way, stale values don't linger around.

That's by the way something that other collectors still suffer from and should be fixed (see https://github.com/prometheus/node_exporter/issues/61 and https://github.com/prometheus/node_exporter/issues/62).

For an example of ConstMetric usage, check out the example Memstats collector, for example: https://github.com/prometheus/client_golang/blob/master/prometheus/example_memstats_test.go